packages/edit-navigation/src/components/layout/use-navigation-block-with-name.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { BlockControls } from '@wordpress/block-editor';
+import { addFilter, removeFilter } from '@wordpress/hooks';
+import {
+	__experimentalEditInPlaceControl as EditInPlaceControl,
+	ToolbarGroup,
+} from '@wordpress/components';
+
+export default function useNavigationBlockWithName( { menuId } ) {
+	const menu = useSelect( ( select ) => select( 'core' ).getMenu( menuId ), [
+		menuId,
+	] );
+
+	const { saveMenu } = useDispatch( 'core' );
+
+	removeFilter( 'editor.BlockEdit', 'core/edit-navigation/with-menu-name' );
+
+	const withMenuName = createHigherOrderComponent(
+		( BlockEdit ) => ( props ) => {
+			if ( props.name !== 'core/navigation' ) {
+				return <BlockEdit { ...props } />;
+			}
+			return (
+				<>
+					<BlockEdit { ...props } />
+					<BlockControls>
+						<ToolbarGroup>
+							<EditInPlaceControl
+								label={ menu?.name ?? '(untitled menu)' }
+								onChange={ ( value ) => {
+									if ( value === '' ) {
+										value = null;
+									}
+									saveMenu( {
+										...menu,
+										name: value ?? '(untitled menu)',
+									} );
+								} }
+							/>
+						</ToolbarGroup>
+					</BlockControls>
+				</>
+			);
+		},
+		'withMenuName'
+	);
+
+	addFilter(
+		'editor.BlockEdit',
+		'core/edit-navigation/with-menu-name',
+		withMenuName
+	);
```

Comment:
We're calling `removeFilter` and `addFilter` on every single render, which is unnecessary work. We should only call `addFilter` when the component mounts and `removeFilter` when the component unmounts.

In addition, we should provide a reference to the filter as the third argument to `removeFilter`.

Something like this:

```js
useEffect( () => {
	const withMenuName = ...;
	addFilter( 'editor.BlockEdit', 'core/edit-navigation/with-menu-name', withMenuName );
	return () => removeFilter( 'editor.BlockEdit', 'core/edit-navigation/with-menu-name', withMenuName );
} );
```

Reply:
In addition, creating a new high-order component `withMenuName` each render is not ideal either. Putting it inside `useEffect` can also solve that.

@noisysocks's example is missing the second argument of `useEffect` though 😆 , should be `[]` or `[menuId]`.

Reply:
My problem with `useEffect` is that when menuId changes (hence menu changes) the edit in place is still closed around the old menu. I can't figure out how to make it refresh. You can check out [e862a25](https://github.com/WordPress/gutenberg/pull/25343/commits/e862a25055a1c806e0091a4d221e65eb4a19dabe) to test.

Reply:
Setting `deps` to `[ menuId ]` should make the effect refresh.